Web tests

ava1on opened the pull request.

gladnik committed:
Added commons-module and tests for user().goTo(place)

Improved project structure by deletion redundant files

gladnik committed:
Changed pom-files structure according to standard

gladnik committed:
Added auxiliary tests

Commit:
add - first web test

Commit:
Добавлены два новых теста:
 - проверка правильности отображения города в заголовке виджета
 - проверка добавления еще одного виджета на страницу

totallynotkate commented on commons-module/src/test/java/ru/qatools/school/webtests/WeatherWebTest.java:
```diff
@@ -0,0 +1,50 @@
+
+public class WeatherWebTest {
+
+    public static final String MOSCOW = "Moscow";
```
Я понимаю, что kurau зачем-то назвал эту строку MOSCOW, но вообще так лучше не делать, лучше писать CITY, ANY_CITY, например. Сейчас непонятно, почему нужно запрашивать именно виджет с Москвой. Может, если там будет Питер, страница вообще не откроется. Или откроется, но будут другие селекторы, и тесты упадут. А может и неважно, какой там город.

kurau replied:
Я сгенерил и забыл поправить потом, конечно же не важно, что тут именно Москва.

totallynotkate commented on steps-module/src/main/java/ru/qatools/school/steps/websteps/DefaultSteps.java:
```diff
@@ -0,0 +1,52 @@
+                onMainPage().getWeatherWidget().get(0).getWidgetTitle().getPrimaryTitle().getCityName() , is(cityName));
+    }
+
+    @Step("Должны видеть на один виджет больше после добавления")
```
По-моему, логика разделения на шаги немного другая. should-методы проверяют, действительно ли все так, как мы ожидаем. Есть ли у нас на экране виджет, правильное ли название города там указано. А вот реальные действия, которые как-то изменяют состояние страницы, вынесены в отдельные шаги: openMainPageWithCity. Поэтому я думаю, что логичнее было бы вынести шаг с нажатием кнопки в отдельный метод, а этот метод превратить в проверку количества виджетов.

kurau replied:
да, так

kurau commented on steps-module/src/main/java/ru/qatools/school/pages/blocks/widgetblocks/PrimaryTitle.java:
```diff
@@ -0,0 +1,27 @@
+    private WebElement city;
+
+    public String getCityName(){
+        return city.getText();
```
Ага, я уже видел такой метод. Если коротко:
- лучше не плодить таких методов
- достаточно метода, который вернёт вам элемент, а при проверке текста воспользоваться матчером `ru.yandex.qatools.htmlelements.matchers.WebElementMatchers.hasText`

kurau commented on steps-module/src/main/java/ru/qatools/school/pages/blocks/widgetblocks/WidgetTitle.java:
```diff
@@ -0,0 +1,24 @@
+
+    @Name("Город")
+    @FindBy(css = ".card-title__primary")
+    private PrimaryTitle primaryTitle;
```
Это не блок, это просто WebElement.

kurau commented on steps-module/src/main/java/ru/qatools/school/steps/websteps/DefaultSteps.java:
```diff
@@ -0,0 +1,52 @@
+    @Step("Должны видеть в заголовке город, указанный в ссылке")
+    public void shouldSeeRightCityInWidgetsTitle(String cityName){
+        assertThat("Город в заголовке совпадает с указанным в ссылке",
+                onMainPage().getWeatherWidget().get(0).getWidgetTitle().getPrimaryTitle().getCityName() , is(cityName));
```
В названии метода фигурирует слово «ссылка», но в реализации про неё ни слова.
Название этого степа очень похоже на название самого теста.
По факту ты здесь проверяешь, что какой-то элемент содержит нужный текст. 
Попробуй упростить название степа в соответствии с тем, что он делает на самом деле.
Либо второй вариант - сделай общий степ, который на вход принимает произвольный элемент и строчку, с которой он будет сравнивать текст элемента. Можешь воспользоваться матчером `ru.yandex.qatools.htmlelements.matchers.WebElementMatchers.hasText`

kurau commented on steps-module/src/main/java/ru/qatools/school/steps/websteps/DefaultSteps.java:
```diff
@@ -0,0 +1,52 @@
+    public void shouldSeeNewAddedWidget(MainPage page){
+        int numberOfWidgets = page.getWeatherWidget().size();
+        page.getAddNewWidgetButton().click();
+        assertThat("Должны видеть на один виджет больше", page.getWeatherWidget().size(), is(numberOfWidgets+1));
```
Опять же, логика этого шага практически полностью повторяет логику самого теста.
Раздели на отдельные шаги.

Commit:
исправление в соответствии с полученными комментариями:
- удален отдельный блок для заголовка виджета, вместо этого используется
  WebElement
- выделен отдельный шаг для нажатия на кнопку добавления виджета
- переписан тест на добавление виджета
- исправлены названия некоторых шагов и тестов

kurau commented on steps-module/src/main/java/ru/qatools/school/steps/websteps/DefaultSteps.java:
```diff
@@ -0,0 +1,56 @@
+
+    @Step("Нажимаем на кнопку добавления виджета")
+    public void addNewWidgetOnMainPage(){
+        onMainPage().getAddNewWidgetButton().click();
```
В итоге в соседних пулреквестах сошлись на мысли, что можно сделать общий метод клика на элемент, который получает на вход произвольный элемент и кликает по нему внутри. Не хочешь попробовать?
